apps/src/lib/node/ledger/protocol/mod.rs

```diff
@@ -121,68 +155,65 @@ where
     block_gas_meter
         .add_base_transaction_fee(tx_length)
         .map_err(Error::GasError)?;
-    match tx {
-        TxType::Raw(_) => Err(Error::TxTypeError),
-        TxType::Decrypted(DecryptedTx::Decrypted(tx)) => {
-            let verifiers = execute_tx(
-                &tx,
-                storage,
-                block_gas_meter,
-                write_log,
-                vp_wasm_cache,
-                tx_wasm_cache,
-            )?;
+    let verifiers = execute_tx(
+        &tx,
+        storage,
+        block_gas_meter,
+        write_log,
+        vp_wasm_cache,
+        tx_wasm_cache,
+    )?;
 
-            let vps_result = check_vps(
-                &tx,
-                storage,
-                block_gas_meter,
-                write_log,
-                &verifiers,
-                vp_wasm_cache,
-            )?;
+    let vps_result = check_vps(
+        &tx,
+        storage,
+        block_gas_meter,
+        write_log,
+        &verifiers,
+        vp_wasm_cache,
+    )?;
 
-            let gas_used = block_gas_meter
-                .finalize_transaction()
-                .map_err(Error::GasError)?;
-            let initialized_accounts = write_log.get_initialized_accounts();
-            let changed_keys = write_log.get_keys();
-            let ibc_event = write_log.take_ibc_event();
+    let gas_used = block_gas_meter
+        .finalize_transaction()
+        .map_err(Error::GasError)?;
+    let initialized_accounts = write_log.get_initialized_accounts();
+    let changed_keys = write_log.get_keys();
+    let ibc_event = write_log.take_ibc_event();
 
-            Ok(TxResult {
-                gas_used,
-                changed_keys,
-                vps_result,
-                initialized_accounts,
-                ibc_event,
-            })
-        }
-        TxType::Protocol(ProtocolTx {
-            tx:
-                ProtocolTxType::EthereumEvents(ethereum_events::VextDigest {
-                    events,
-                    ..
-                }),
+    Ok(TxResult {
+        gas_used,
+        changed_keys,
+        vps_result,
+        initialized_accounts,
+        ibc_event,
+    })
+}
+
+// Apply a derived transaction to storage based on some protocol transaction.
+// The logic here must be completely deterministic and will be executed by all
+// full nodes every time a protocol transaction is included in a block. Storage
+// is updated natively rather than via the wasm environment, so gas does not
+// need to be charged and validity predicates are bypassed. A [`TxResult`]
+// containing changed keys and the like should be returned in the normal way.
+pub(crate) fn apply_protocol_tx<'a, D, H>(
+    tx: ProtocolTxType,
+    _storage: &'a mut Storage<D, H>,
```

Comment:
This may eventually be tightened further to be an impl trait of some subset of `impl Storage`. So far for processing Ethereum events, only `storage.read`, `storage.write` and `storage.has_key` has been needed.

Reply:
add a TODO there, maybe?